### ISSUE TITLE
Random riak.client.multiget.MultiGetPool - threading.py Exception [JIRA: CLIENTS-817]

### ISSUE THREAD

**Issue by braoru** [basho/riak-python-client#449]
Hello

I create small tools to manipulate riak data, when I use python 3.4.3 I commonly get random threading error (1 time on 5 or 6).

```Bash
Exception ignored in: <bound method MultiGetPool.__del__ of <riak.client.multiget.MultiGetPool object at 0x7f64651e0ac8>>
Traceback (most recent call last):
  File "/home/sebastien/Documents/dev/components-tools/lib64/python3.4/site-packages/riak/client/multiget.py", line 123, in __del__
  File "/home/sebastien/Documents/dev/components-tools/lib64/python3.4/site-packages/riak/client/multiget.py", line 110, in stop
  File "/usr/lib64/python3.4/threading.py", line 517, in set
  File "/usr/lib64/python3.4/threading.py", line 359, in notify_all
  File "/usr/lib64/python3.4/threading.py", line 342, in notify
TypeError: 'NoneType' object is not callable

```

It's seem to be related to a try to stop an empty pool but not sure.

Python version and Riak lib version

```Bash
(components-tools) [sebastien@aspros components-tools]$ python --version
Python 3.4.3
(components-tools) [sebastien@aspros components-tools]$ pip show riak
---
Metadata-Version: 1.1
Name: riak
Version: 2.4.2
Summary: Python client for Riak
Home-page: https://github.com/basho/riak-python-client
Author: Basho Technologies
Author-email: clients@basho.com
License: Apache 2
Location: /home/sebastien/Documents/dev/components-tools/lib/python3.4/site-packages
Requires: six, python3-protobuf
Classifiers:
  License :: OSI Approved :: Apache Software License
  Intended Audience :: Developers
  Operating System :: OS Independent
  Programming Language :: Python :: 2.7
  Programming Language :: Python :: 3.3
  Programming Language :: Python :: 3.4
  Programming Language :: Python :: 3.5
  Topic :: Database
(components-tools) 
```

**Comment by lukebakken** [comment]
Could you please provide code to reproduce this?

**Comment by braoru** [comment]
Hello

I can separate the code in a dedicated snippet

https://gist.github.com/braoru/82099448b1ca2cc348fe2aea6a47a8f9

Then I can't reproduce the error in the snippet. The exact same code in the original function does generate the error. 

The function itself is not a lot more complicated https://gist.github.com/braoru/5b16b395787e393024b97f14d537bca7

On 300 run I get 13 run with the ignored Exception.

**Comment by braoru** [comment]
We are using local bucket set as index. Here it's a simple index but sometime we use them as pre-filter to save cpu within riak. With secondary index / yokozuma we add serious CPU intensive trouble.

And now we are only using bitcask with ttl bucket then we are using only local self-made index.

**Comment by evrenesat** [comment]
We also getting same errors randomly while testing our project without any reproducible cause.

```python

platform darwin -- Python 3.4.2 -- py-1.4.28 -- pytest-2.7.1 
...
========================== 41 passed in 58.19 seconds ==========================
Exception ignored in: <bound method MultiGetPool.__del__ of <riak.client.multiget.MultiGetPool object at 0x102648ef0>>
Traceback (most recent call last):
  File "/Users/evren/.virtualenvs/pyoko3/src/riak/riak/client/multiget.py", line 123, in __del__
  File "/Users/evren/.virtualenvs/pyoko3/src/riak/riak/client/multiget.py", line 110, in stop
  File "/usr/local/Cellar/python3/3.4.2/Frameworks/Python.framework/Versions/3.4/lib/python3.4/threading.py", line 517, in set
  File "/usr/local/Cellar/python3/3.4.2/Frameworks/Python.framework/Versions/3.4/lib/python3.4/threading.py", line 359, in notify_all
  File "/usr/local/Cellar/python3/3.4.2/Frameworks/Python.framework/Versions/3.4/lib/python3.4/threading.py", line 342, in notify
TypeError: 'NoneType' object is not callable

Process finished with exit code 0

```

**Comment by fifi147** [comment]
I also have these issues after switching from Python 2.7 to Python 3.4. Also I could not localize the exact cause.

**Comment by lukebakken** [comment]
Just to confirm with everyone who has commented, this issue only appears when the Python process stops or you are shutting down the client?

**Comment by braoru** [comment]
Hello, it's seem to appear when the client is shutting down.

I will run more test on Monday and come back!

**Comment by evrenesat** [comment]
I see it only at the end of our tests, so it seems it appears when the process stops.

**Comment by fifi147** [comment]
Hello, is there any update on this? Its the only thing that stops me from switching my module using Riak to 3.4 :)

**Comment by lukebakken** [comment]
@fifi147 - since this is a random issue that happens on client shutdown, it is low priority to fix at the moment. A PR with a fix would be welcome.

**Comment by linkdd** [comment]
Hello,

After searching a while, it seems that the ``riak.client.multi.MultiPool.__del__()`` method is not called when the error is **not** raised.

This is probably due to some circular references Python cannot solve so he is just not destroying the object.

When the threads aren't finished, they are still holding the reference to the ``MultiPool`` object (because the target is the method ``MultiPool._worker_method``).

The weird thing is that when the destructor is called, the ``threading`` module have in ``globals()`` everything that is not built-in set to ``None`` (probably due to Python exiting?).

When I'm closing a connection, I manually have to stop the pools:

```python
if client._multiget_pool:
    client._multiget_pool.stop()

if client._multiput_pool:
    client._multiput_pool.stop()

client.close()
```

In order to avoid the method ``stop()`` to be called again when the property gets deleted, I modified the ``riak.client.multi.MultiPool.stop()`` method:

```python
    def stop(self):
        """
        Signals the worker threads to exit and waits on them.
        """
        if not self.stopped():  # make sure the pool isn't already stopped
            self._stop.set()
            for worker in self._workers:
                worker.join()
```

And finally, I had to manually close the two pools instantiated in the module, in order to avoid the error at the exit of my program (I should use ``sys.atexit()`` for this):

```python
from riak.client.multi import RIAK_MULTIGET_POOL, RIAK_MULTIPUT_POOL

RIAK_MULTIGET_POOL.stop()
RIAK_MULTIPUT_POOL.stop()
```

Those are not bugfixes, just workarounds.
